internal/metamorph/store/postgresql/postgres.go

```diff
@@ -573,6 +591,135 @@ func (p *PostgreSQL) UpdateStatusBulk(ctx context.Context, updates []store.Updat
 	return res, nil
 }
 
+func (p *PostgreSQL) UpdateDoubleSpend(ctx context.Context, updates []store.UpdateStatus) ([]*store.StoreData, error) {
+	qBulk := `
+		UPDATE metamorph.transactions
+			SET
+			status=bulk_query.status,
+			reject_reason=bulk_query.reject_reason,
+			competing_txs=bulk_query.competing_txs
+			FROM
+			(
+				SELECT * FROM UNNEST($1::BYTEA[], $2::INT[], $3::TEXT[], $4::TEXT[])
+				AS t(hash, status, reject_reason, competing_txs)
+			) AS bulk_query
+			WHERE metamorph.transactions.hash=bulk_query.hash
+				AND metamorph.transactions.status <= bulk_query.status
+				AND LENGTH(metamorph.transactions.competing_txs) < LENGTH(bulk_query.competing_txs)
+		RETURNING metamorph.transactions.stored_at
+		,metamorph.transactions.announced_at
+		,metamorph.transactions.mined_at
+		,metamorph.transactions.hash
+		,metamorph.transactions.status
+		,metamorph.transactions.block_height
+		,metamorph.transactions.block_hash
+		,metamorph.transactions.callback_url
+		,metamorph.transactions.callback_token
+		,metamorph.transactions.full_status_updates
+		,metamorph.transactions.reject_reason
+		,metamorph.transactions.competing_txs
+		,metamorph.transactions.raw_tx
+		,metamorph.transactions.locked_by
+		,metamorph.transactions.merkle_path
+		,metamorph.transactions.retries
+		;
+    `
+
+	txHashes := make([][]byte, len(updates))
+	for i, update := range updates {
+		txHashes[i] = update.Hash.CloneBytes()
+	}
+
+	tx, err := p.db.Begin()
+	if err != nil {
+		return nil, err
+	}
+
+	// Get current competing transactions and lock them for update
+	rows, err := tx.QueryContext(ctx, `SELECT hash, competing_txs FROM metamorph.transactions WHERE hash in (SELECT UNNEST($1::BYTEA[])) FOR UPDATE`, pq.Array(txHashes))
+	if err != nil {
+		if err := tx.Rollback(); err != nil {
+			return nil, err
+		}
+		return nil, err
+	}
+	defer rows.Close()
+
+	dbData := make([]*store.StoreData, 0)
+	for rows.Next() {
+		data := &store.StoreData{}
+
+		var hash []byte
+		var competingTxs string
+
+		err := rows.Scan(
+			&hash,
+			&competingTxs,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		if len(hash) > 0 {
+			data.Hash, err = chainhash.NewHash(hash)
```

Comment:
Maybe it would be overengineering, premature microoptimization, or even an incorrect solution, but...

BUT maybe we don't need to use `chainhash`es here and then compare them later. We already have pure bytes of hashes (from line 630),  so maybe it would be simpler to compare them directly?

Reply:
I think it's a brilliant idea, I implemented it in the next PR: https://github.com/bitcoin-sv/arc/pull/527

It was easier for me that way, because I already had a separate function for this. Please check it there :)